File: 70-climbing-stairs/climbing-stairs.java
class Solution {
    // private int[] dp = new int[46];
    // private int helper(int i,int n){
    //     if(i==n-1){
    //         return 1;
    //     }
    //     if(i==n-2){
    //         return 2;
    //     }

    //     if(dp[i]!=-1){
    //         return dp[i];
    //     }

    //     return dp[i] = helper(i+1,n) + helper(i+2,n);
    // }
    public int climbStairs(int n) {
        // Arrays.fill(dp,-1);
        // return helper(0,n);

        if(n==0 || n==1){
            return n;
        }

        // int[] dp = new int[n+1];
        // Arrays.fill(dp,0);

        int prev2 = 1;
        int prev = 2;

        for(int i = n-3;i>=0;i--){
            int curr = prev + prev2;
            prev2 = prev;
            prev = curr;
        }

        return prev;
    }
}

File: 70-climbing-stairs/climbing-stairs.cpp
class Solution {
    private:
    int memo[46];
    int solve(int curr, int n){
        if(curr>n){
            return 0;
        }

        if(curr==n){
            return 1;
        }

        if(memo[curr]!=-1){
            return memo[curr];
        }

        int oneStep = solve(curr+1, n);
        int twoSteps = solve(curr+2, n);

        return memo[curr] = oneStep + twoSteps;
    }
public:
    int climbStairs(int n) {
        memset(memo, -1 , sizeof(memo));
        return solve(0, n);
    }
};